src/QirRuntime/lib/Tracer/README.md

```diff
@@ -0,0 +1,186 @@
+4. the logic for frame tracking;
+5. output of the collected statistics;
+6. (_lower priority_) the scheduling component to optimize depth and/or width of the circuit.
+
+## Layering ##
+
+_Definition_: ___Time___ is an integer-valued function on all quantum operations in a program (gates, measurements,
+ qubits allocation/release). For each gate there are start and end times. For each qubit, there are times when the qubit
+ is allocated and released. Start time of a gate cannot be less than allocation time of any of the qubits the gate uses.
+ If two gates or measurements use the same qubit, one of the gates must have start time greater than or equal to the end
+ time of the other.
+
+A sequentially executed quantum program can be assigned a trivial time function, when all quantum operations have
+ duration of 1 and unique start times, ordered to match the flow of the program. Layering compresses the timeline by
+ assuming that some operations might be executed simultaneously while allowing for different operations to have various
+ durations.
+
+_Definition_: Provided a valid _time_ function for the program a ___layer of duration N at time T, denoted as L(T,N),___
+ is a subset of operations in the program such that all of these operations have start time greater or equal _T_ and
+ finish time less than _T + N_. The program is ___layered___ if all gates in it are partitioned into layers, that don't
+ overlap in time. The union of all qubits that are involved in operations of a given layer, will be denoted _Qubits(T,N)_.
+
+A sequential program can be trivially layered such that each layer contains exactly one operation. Notice, that the
+ definition of layer doesn't require the gates to be executed _in parallel_. For example, all gates in a fully sequential
+ program can be also placed into a single layer L(0, infinity). Some gates might be considered to be very cheap and take
+ zero time to execute, those gates can be added to a layer even if they act on the same qubit another gate in this layer
+ is acting on and have to be executed sequentially within the layer.
+
+### The Resource Tracer's Layering Algorithm ###
+
+As the tracer is executing a sequential quantum program, it will compute a time function and corresponding layering
+ using the _conceptual_ algorithm, described below (aka "tetris algorithm"). The actual implementation of layering might
+ be done differently, as long as the resulting layering is the same as if running the conceptual algorithm.
+
+_Definition_: A ___barrier___ is a layer that no more operations can be added into.
+
+A user will be able to inject global barriers by calling `__quantum__qis__global_barrier` function. The user can choose
+ duration of a barrier which would affect start time of the following layers but no operations will be added to a barrier,
+ independent of its width.
+
+#### The conceptual algorithm ####
+
+1. The tracer must be set the preferred layer duration: P.

```

Comment:
> The tracer **must** be set the preferred layer duration: P

The word "**must**" tells me that there cannot be _default_ preferred layer duration.
Did you means something like "**can be**" (instead of "must"), which would tell me that if it is not set then the default will be used.

Reply:
=> "Note: The tracer assumes that the preferred layer duration is _P_." I need to introduce notation for the default layer duration so I can use it in the following steps of the algorithm.